### PR TITLE
Docker support

### PR DESCRIPTION
**Pull request by craig-rueda**
### CATEGORY

Choose one

- [ ] Bug Fix
- [x] Enhancement (new features, refinement)
- [ ] Refactor
- [ ] Add tests
- [ ] Build / Development Environment
- [ ] Documentation

### What:
* Updates the way Superset deals with Docker.
* Updates the local docker-compose setup
* Moves docker stuff out of `contrib` and into `docker`, making it a little more "first-class"
* Dockerfile is now more production ready
* Splits development work into several helper containers (init, node, worker) as opposed to one uber container
* Removes node from final image
* Builds in a multi-stage way in order to minimize final size

### Why:
* We would like to start building/pushing prod-ready Docker images that will mirror the Apache release cycles

### What's next:
* CI will be hooked up to start building the `Dockerfile` and then push it up somewhere that it can be used by everyone

### CODE REVIEW

File: docker/pythonpath_dev/superset_config.py [hunk 1]
```diff
@@ -14,9 +14,21 @@
...
+# development environments. Also note that superset_config_docker.py is imported
+# as a final step as a means to override "defaults" configured here
+#
+
+import logging
 import os
 
 
+logger = logging.getLogger()
+
```
Comment (dpgaspar) [inline docker/pythonpath_dev/superset_config.py]: nit: Can you change single quotes to double quotes on this file

File: docker/docker-entrypoint.sh [hunk 2]
```diff
@@ -15,24 +15,15 @@
...
-        --workers $((2 * $(getconf _NPROCESSORS_ONLN) + 1)) \
+if [ "${#}" -ne 0 ]; then
+    exec "${@}"
+else
+    gunicorn --bind  "0.0.0.0:${SUPERSET_PORT}" \
+        --workers 1 \
         --timeout 60 \
         --limit-request-line 0 \
         --limit-request-field_size 0 \
         "superset.app:create_app()"
```
Comment (dpgaspar) [inline docker/docker-entrypoint.sh]: Replace `superset.app:create_app()` by `${FLASK_APP}`

File: docker/pythonpath_dev/superset_config_local.example [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+#
+# This is an example "local" configuration file. In order to set/override config
+# options that ONLY apply to your local environment, simply copy/rename this file
+# to docker/pythonpath/superset_config_docker.py
+# It ends up being imported by docker/superset_config.py which is loaded by
+# superset/config.py
+#
+
+SQLALCHEMY_DATABASE_URI = 'postgresql+psycopg2://pguser:pgpwd@some.host/superset'
```
Comment (dpgaspar) [inline docker/pythonpath_dev/superset_config_local.example]: nit: double quotes

File: Dockerfile [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+            default-libmysqlclient-dev \
+            libpq-dev \
+        && rm -rf /var/lib/apt/lists/*
+
+COPY requirements* ./setup.py ./MANIFEST.in ./README.md ./app/
+COPY superset /app/superset
+
+RUN cd /app \
+        && pip install --upgrade setuptools pip \
+        && pip install -r requirements.txt -r requirements-dev.txt \
```
Comment (dpgaspar) [inline Dockerfile]: I understand the need, but I really don't like this `requirements-dev` here, contains lot's of things that are not needed and his changed by contributions taking into context a development environment not an official docker build. We could use a `requirements-extra` for this.

**Review by dpgaspar — COMMENTED** [review]
Looks good, left some comments

**Comment by craig-rueda** [comment]
@dpgaspar - addressed all of your comments. Thanks for the review!

**Review by willbarrett — APPROVED** [review]

**Review by dpgaspar — APPROVED** [review]

**Comment by jackyourkill** [comment]
Please update the documentation at https://superset.incubator.apache.org/installation.html, since this breaks the steps for Start with Docker

**Comment by craig-rueda** [comment]
Done in #8735 😄

**Pull request merged**